File: src/python/algorithms/strings/removeAllAdjacentDuplicates/remove_all_adjacent_duplicates.py
# Stack with (char, count) pairs  TC: O(n) SC: O(n)
def remove_duplicates(s, k):
    stack = []  # each entry: [char, count]

    for ch in s:
        if stack and stack[-1][0] == ch:
            stack[-1][1] += 1
            if stack[-1][1] == k:
                stack.pop()
        else:
            stack.append([ch, 1])

    return "".join(ch * count for ch, count in stack)


if __name__ == "__main__":
    test_cases = [
        ("abcd",                   2, "abcd"),
        ("deeedbbcccbdaa",         3, "aa"),
        ("pbbcggttciiippooaais",   2, "ps"),
        ("aabbaa",                 2, ""),
        ("abcddcba",               2, ""),
    ]

    for s, k, expected in test_cases:
        result = remove_duplicates(s, k)
        status = "PASS" if result == expected else "FAIL"
        print(f"[{status}] result={result} expected={expected}")

File: src/javascript/algorithms/strings/removeAllAdjacentDuplicates/removeAllAdjacentDuplicates.js
// Stack with (char, count) pairs  TC: O(n) SC: O(n)
function removeDuplicates(s, k) {
    const stack = []; // each entry: { ch, count }

    for (const ch of s) {
        if (stack.length > 0 && stack[stack.length - 1].ch === ch) {
            stack[stack.length - 1].count++;
            if (stack[stack.length - 1].count === k) {
                stack.pop();
            }
        } else {
            stack.push({ ch, count: 1 });
        }
    }

    return stack.map(({ ch, count }) => ch.repeat(count)).join("");
}

function runTests() {
    const tests = [
        { s: "abcd",                   k: 2, expected: "abcd" },
        { s: "deeedbbcccbdaa",         k: 3, expected: "aa" },
        { s: "pbbcggttciiippooaais",   k: 2, expected: "ps" },
        { s: "aabbaa",                 k: 2, expected: "" },
        { s: "abcddcba",               k: 2, expected: "" },
    ];

    tests.forEach(({ s, k, expected }) => {
        const result = removeDuplicates(s, k);
        const status = result === expected ? "PASS" : "FAIL";
        console.log(`[${status}] result=${result} expected=${expected}`);
    });
}

runTests();
